compiler/rustc_ty_utils/src/opaque_types.rs

```diff
@@ -35,33 +52,91 @@ impl<'tcx> OpaqueTypeCollector<'tcx> {
-            DefKind::AssocFn | DefKind::AssocTy | DefKind::AssocConst => {
-                Some(self.tcx.local_parent(self.item))
-            }
-            other => span_bug!(
-                self.tcx.def_span(self.item),
-                "unhandled item with opaque types: {other:?}"
-            ),
+    fn normalize_if_possible<T: TypeFoldable<TyCtxt<'tcx>>>(
+        &self,
+        span: Span,
+        value: T,
+    ) -> Result<T, ErrorGuaranteed> {
+        if !value.has_projections() {
+            return Ok(value);
         }
+
+        let infcx = self.tcx.infer_ctxt().build();
+        let param_env = self.tcx.param_env(self.item);
+
+        with_replaced_escaping_bound_vars(
+            &infcx,
+            &mut vec![None; self.binder.as_usize()],
+            value,
+            |value| {
+                let mut fulfill_cx = <dyn TraitEngine<'tcx>>::new(&infcx);
+
+                let normalized_value = match infcx
+                    .at(&ObligationCause::misc(span, self.item), param_env)
+                    .deeply_normalize(value, &mut *fulfill_cx)
+                {
+                    Ok(t) => t,
+                    Err(errors) => {
+                        return Err(self
+                            .tcx
+                            .sess
+                            .delay_span_bug(span, format!("{errors:#?} in {:?}", self.item)));
+                    }
+                };
+
+                let InferOk { value: implied_wf_types, obligations } =
+                    infcx.at(&ObligationCause::misc(span, self.item), param_env).normalize(
+                        self.tcx
+                            .assumed_wf_types(self.item)
+                            .iter()
+                            .map(|(ty, _span)| *ty)
+                            .collect::<Vec<_>>(),
+                    );
+                fulfill_cx.register_predicate_obligations(&infcx, obligations);
+
+                let errors = fulfill_cx.select_all_or_error(&infcx);
+                if !errors.is_empty() {
+                    return Err(self.tcx.sess.delay_span_bug(span, format!("{errors:#?}")));
+                }
+
+                let outlives_env = OutlivesEnvironment::with_bounds(
+                    param_env,
+                    infcx.implied_bounds_tys(
+                        param_env,
+                        self.item,
+                        implied_wf_types.into_iter().collect(),
+                    ),
+                );
+                let errors = infcx.resolve_regions(&outlives_env);
+                if !errors.is_empty() {
+                    return Err(self.tcx.sess.delay_span_bug(span, format!("{errors:#?}")));
+                }
+
+                let resolved_value = match infcx.fully_resolve(normalized_value) {
+                    Ok(resolved_value) => resolved_value,
+                    Err(f) => {
+                        return Err(self.tcx.sess.delay_span_bug(span, format!("{f:?}")));
```

Comment:
I think we're lacking the test coverage to hit those delay span bugs. Most of our TAIT tests fail to compile. You'd need to write a passing test that hits this code path on an unrelated function. It should not error or fail to compile, because it's likely only failing because stuff is too generic.

Reply:
Well, we should never hit these delay span bugs, except for when we have malformed param-envs. I'm not certain we can actually hit these code paths.